src/Protocols/DfLike.hs

```diff
@@ -41,931 +68,1546 @@ module Protocols.DfLike
-hasPayload dfA = Maybe.isJust . getPayload dfA
-{-# INLINE hasPayload #-}
-
-mapPayload ::
-  forall df dom a b.
-  ( DfLike dom df a
-  , DfLike dom df b ) =>
-  Proxy (df a) ->
-  Proxy (df b) ->
-  (Payload a -> Payload b) ->
-  Data df a ->
-  Data df b
-mapPayload dfA dfB f fwdA =
-  setPayload dfA dfB fwdA (fmap f (getPayload dfA fwdA))
-{-# INLINE mapPayload #-}
-
-
--- | Allows conversion of one DfLike protocol into another, given two functions that
--- transform the data and ack type of one protocol to the other.
-mapDfLike ::
-  forall df1 df2 dom a1 a2.
-  ( DfLike dom df1 a1
-  , DfLike dom df2 a2 ) =>
-  Proxy (df1 a1) ->
-  Proxy (df2 a2) ->
-  (Data df1 a1 -> Data df2 a2) ->
-  (Ack df2 a2 -> Ack df1 a1) ->
-  Circuit (df1 a1) (df2 a2)
-mapDfLike Proxy Proxy fData fAck = Circuit (CE.unbundle . fmap go . CE.bundle)
-  where
-    go :: (Data df1 a1, Ack df2 a2) -> (Ack df1 a1, Data df2 a2)
-    go (d, ack) = (fAck ack, fData d)
+  type Dom
+    (Reverse (Axi4ReadAddress dom confAR userAR),
+     Axi4ReadData dom confR userR dat)
+     = dom
+  type BwdPayload -- sent at the beginning of each burst (NOT each transfer)
+    (Reverse (Axi4ReadAddress dom confAR userAR),
+     Axi4ReadData dom confR userR dat)
+     = Axi4ReadAddressInfo confAR userAR
+  type FwdPayload
+    (Reverse (Axi4ReadAddress dom confAR userAR),
+     Axi4ReadData dom confR userR dat)
+     = (dat, userR, ResponseType (RKeepResponse confR))
+
+  toDfCircuit _ = toDfCircuitHelper s0 blankOtp stateFn where
+    -- ( burst len left in transfer
+    -- , read id currently replying to )
+    s0 = (0, errorX "DfLike for Axi4: No initial value for read id" )
+
+    blankOtp = (S2M_ReadAddress { _arready = False }, S2M_NoReadData)
+
+    stateFn (addrVal, dataAck) dfAckIn dfDatIn = do
+      (addrAck, dfDatOut) <- processAddr addrVal dfAckIn
+      (dataVal, dfAckOut) <- sendData dfDatIn dataAck
+      P.pure ((addrAck,dataVal),dfDatOut,dfAckOut)
+
+    processAddr M2S_NoReadAddress _ = P.pure (S2M_ReadAddress { _arready = False }, Nothing)
+    processAddr msg dfAckIn = do
+      (burstLenLeft,_) <- get
+      when (burstLenLeft == 0) $ put (succResizing $ fromMaybe 0 (fromKeepType $ _arlen msg), _arid msg)
+      P.pure (S2M_ReadAddress{ _arready = burstLenLeft == 0 && dfAckIn }, Just (axi4ReadAddrMsgToReadAddrInfo msg))
+
+    succResizing :: (KnownNat n) => Index n -> Index (n+1)
+    succResizing n = (resize n) + 1
+
+    sendData dfDatIn dataAck = do
+      (burstLenLeft,readId) <- get
+      case (burstLenLeft == 0, dfDatIn) of
+        (False, Just (_rdata, _ruser, _rresp)) -> do
+          put (burstLenLeft-1, readId)
+          P.pure
+            ( S2M_ReadData
+              { _rid = readId
+              , _rlast = burstLenLeft == 1
+              , _rdata, _ruser, _rresp }
+            , _rready dataAck)
+        _ -> P.pure (S2M_NoReadData, False)
+
+instance
+  ( GoodAxi4ReadAddressConfig confAR
```

Comment:
Bit late, but I think we should rename this to `Axi4ReadAddressConfigConstraints`. That can go in a different PR.